File: product/utils/meta.go
package utils

import (
	"os"
	"strings"
	"time"

	"github.com/autobar-dev/services/product/types"
)

func GetMetaFactors() *types.MetaFactors {
	commit_sha := ""
	version := ""

	if commit_sha_bytes, err := os.ReadFile(".meta/HASH"); err == nil {
		commit_sha = strings.TrimSpace(string(commit_sha_bytes))
	}

	if version_bytes, err := os.ReadFile(".meta/VERSION"); err == nil {
		version = strings.TrimSpace(string(version_bytes))
	}

	return &types.MetaFactors{
		StartTime: time.Now(),
		Hash:      commit_sha,
		Version:   version,
	}
}

func GetMetaFromFactors(mf types.MetaFactors) *types.Meta {
	now := time.Now()
	uptime := int64(now.Sub(mf.StartTime).Seconds())

	return &types.Meta{
		Uptime:  uptime,
		Hash:    mf.Hash,
		Version: mf.Version,
	}
}

File: emailtemplate/src/utils/meta.ts
import { readFile } from "fs/promises"; 
import { join } from "path";
import { Meta, MetaFactors } from "../types/meta";

export async function getMetaFactors(): Promise<MetaFactors> {
  let commit_sha = "";
  try {
    const commit_sha_bytes = await readFile(join(__dirname, "..", "..", ".meta", "HASH"));
    commit_sha = commit_sha_bytes.toString("utf8");
  } catch (_e) {}

  let version = "";
  try {
    const version_bytes = await readFile(join(__dirname, "..", "..", ".meta", "VERSION"));
    version = version_bytes.toString("utf8");
  } catch (_e) {}

  commit_sha = commit_sha.trim() || "";
  version = version.trim() || "";
  
  return {
    startTime: new Date(),
    hash: commit_sha,
    version,
  } satisfies MetaFactors;
}

export function getMetaFromFactors(metaFactors: MetaFactors): Meta {
  return {
    uptime: Math.floor((new Date().getTime() - metaFactors.startTime.getTime()) / 1000),
    hash: metaFactors.hash,
    version: metaFactors.version,
  };
}
